### PR TITLE
Set the LastModified header for raw files

### LINKED ISSUE #18354: Add Last-Modified Header equal to the lastcommitdate of a file.
### Feature Description

In use cases where you may need a simple script (bash, python, etc) to be the current version, it would be helpful for the Gitea server to provide the "Last-Modified" header set to the lastcommitdate. 

This would allow a script to 'self check' the headers on any direct file link, and perform a local check to see if the gitea version is of a newer version than the local copy.  

This is especially useful where you may not be able (or allowed) to install `git` to run on a system. 

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by zeripath**
Although the use of LastModified dates for caching of git objects should be
discouraged (as it is not native to git - and there are a LOT of ways this
could be incorrect) - LastModified dates can be a helpful somewhat more human
way of caching for simple cases.

This PR adds this header and handles the If-Modified-Since header to the /raw/
routes for branch, tag and commit.

Fix #18354

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: routers/web/repo/download.go [hunk 1]
```diff
@@ -79,34 +82,57 @@ func ServeBlobOrLFS(ctx *context.Context, blob *git.Blob) error {
 	return common.ServeBlob(ctx, blob)
 }
 
+func getBlobForEntry(ctx *context.Context) *git.Blob {
+	entry, err := ctx.Repo.Commit.GetTreeEntryByPath(ctx.Repo.TreePath)
+	if err != nil {
+		ctx.ServerError("GetBlobByPath", err)
```
Comment (Gusted) [inline routers/web/repo/download.go]:
```suggestion
		ctx.ServerError("GetTreeEntryByPath", err)
```

File: routers/web/repo/download.go [hunk 2]
```diff
@@ -79,34 +82,57 @@ func ServeBlobOrLFS(ctx *context.Context, blob *git.Blob) error {
...
+		return nil
+	}
+
+	if entry.IsDir() || entry.IsSubModule() {
```
Comment (Gusted) [inline routers/web/repo/download.go:97]: Merge these 2 conditionals to line 87?

**Review by Gusted — COMMENTED** [review]
Hmm, if this is going to cause _major_ problems, we should revert this PR. But from the code it's looking good, it should send a new copy of the file when CommitterDate is changed, which should be sufficient to have good caching.

File: routers/web/repo/download.go (hunk 2 above)
Comment (zeripath) [reply routers/web/repo/download.go:97]: One is a ServerError but this is a NotFound.

**Comment by zeripath** [comment]
It's a very poor cache and could lead to things being cached when they shouldn't be - for example if there has been a force-push.

However if you want to use LastModified instead of the etag then this is the risk you take.

I guess we need to ask what @silverwind thinks - my only worry is that Browsers start sending both and the IfModifiedSince incorrectly detects no change when there is one.

**Comment by Gusted** [comment]
>my only worry is that Browsers start sending both and the IfModifiedSince incorrectly detects no change when there is one.

What do you mean sending both?

**Comment by zeripath** [comment]
I mean an HTTP Request sends both `If-Modified-Since` and `ETag`.

**Comment by zeripath** [comment]
Actually I guess this just means we should deal with both headers at the same time.

**Comment by Gusted** [comment]
> I mean an HTTP Request sends both `If-Modified-Since` and `If-None-Match`.

```
When used in combination with If-Modified-Since, If-None-Match has precedence (if the server supports it).
```

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match

Official ref: https://httpwg.org/specs/rfc7232.html#precedence

**Comment by silverwind** [comment]
Browsers by default send none of the headers. If they encounter `Etag`, they will send `If-None-Match` on subsequent requests, and if they see `Last-Modified`, they will send `If-Modified-Since`, assuming no other directives like `no-store` prevent caching.

**Comment by zeripath** [comment]
So if we send both `Last-Modified` and `Etag` they will send both `If-Modified-Since` and `If-None-Match`?

If so this code is likely incorrect and we need to make it so that we only pay attention to If-Modified-Since iff `If-None-Match`

**Review by Gusted — APPROVED** [review]
LGTM

**Review by singuliere — APPROVED** [review]

**Comment by wxiaoguang** [comment]
How about work on https://github.com/go-gitea/gitea/pull/18448 ? IIRC `http.ServeContent` can handle these headers correctly.

File: routers/web/repo/download.go [hunk 3]
```diff
@@ -122,7 +150,7 @@ func DownloadByID(ctx *context.Context) {
 		}
 		return
 	}
-	if err = common.ServeBlob(ctx, blob); err != nil {
+	if err = common.ServeBlob(ctx, blob, time.Time{}); err != nil {
```
Comment (6543) [inline routers/web/repo/download.go:157]: do blobs dont have a timestamp?

Comment (6543) [reply routers/web/repo/download.go:157]: they dont :/

**Review by 6543 — APPROVED** [review]

**Review by 6543 — APPROVED** [review]

**Comment by 6543** [comment]
:rocket:

**Pull request merged**